### PR TITLE
Add billing for Odoo 16

### PR DESCRIPTION
**Pull request by Kidswiss**
The whole billing for vshn services consists of two parts:

* prometheus rules deployed on each cluster with billable AppCat service that aggregate any deployed instance into our custom metric
* cronjobs to read from mimir and then push to odoo 16, these will only be deployed to the appuio cloud prod cluster

## Checklist

- [x] The PR has a meaningful title. It will be used to auto generate the
      changelog.
      The PR has a meaningful description that sums up the change. It will be
      linked in the changelog.
- [x] PR contains a single logical change (to build a better changelog).
- [x] Categorize the PR by adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.
- [x] Link this PR to related issues or PRs.

### CODE REVIEW

File: component/class/defaults.yml [hunk 1]
```diff
@@ -65,8 +69,49 @@ parameters:
 
     appuioManaged: true
     tenantID: ${cluster:tenant}
+    salesOrderID: TBD
```
Comment (TheBigLee) [inline component/class/defaults.yml:72]: Why do we have this twice? Once here and once under `billing`

File: component/class/defaults.yml [hunk 2]
```diff
@@ -65,8 +69,49 @@ parameters:
...
     quotasEnabled: false
     grpcEndpoint: host.docker.internal:9443
+
+    billing:
+      namespace: ${appcat:namespace}
+      clusterId: ${cluster:name}
```
Comment (TheBigLee) [inline component/class/defaults.yml:78]: The `tenantID` is outside of `billing` and the `clusterId` inside. This is inconsistent (also once the ID is all capital and once camel case.

File: component/component/billing.jsonnet [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+local netPol = import 'billing_netpol.libsonnet';
+local kap = import 'lib/kapitan.libjsonnet';
+local kube = import 'lib/kube.libjsonnet';
+local inv = kap.inventory();
+// The hiera parameters for the component
+local params = inv.parameters.appcat;
+
+local formatImage = function(ref) '%(registry)s/%(repository)s:%(tag)s' % ref;
+
+local escape = function(str)
```
Comment (TheBigLee) [inline component/component/billing.jsonnet:13]: A few comments to explain what this function does would be nice  (yes the name should be already a hint, but might still not be clear to everyone)

File: component/component/billing.jsonnet [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    },
+  };
+
+local generateCloudAndManaged = function(name)
+
+  // For postgresql we have a missmatch between the label and the name in our definition.
+  local queryName = if name == 'postgres' then name + 'ql' else name;
+
+  local managedQuery = 'appcat:metering{label_appuio_io_billing_name="appcat-' + queryName + '",label_appcat_vshn_io_sla="%s"}';
+  local cloudQuery = managedQuery + ' * on(label_appuio_io_organization) group_left(sales_order_id) label_replace(appuio_control_organization_info, "label_appuio_io_organization", "$1", "name", "(.*)")';
```
Comment (TheBigLee) [inline component/component/billing.jsonnet:154]:
```suggestion
  local cloudQuery = managedQuery + ' * on(label_appuio_io_organization) group_left(sales_order) label_replace(appuio_control_organization_info, "label_appuio_io_organization", "$1", "name", "(.*)")';
```

File: component/component/billing_cronjob.libsonnet [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  'app.kubernetes.io/managed-by': 'commodore',
+  'app.kubernetes.io/part-of': 'syn',
+};
+
+local cronJob = function(name, scheduleName, jobSpec)
+  local truncatedName = if std.length(name) < 52 then
+    name
+  else
+    '%s-%s' % [ std.substr(name, 0, 36), std.substr(std.md5(name), 0, 15) ];
+  local schedule = '30 * * * *';  //params.schedules[scheduleName];
```
Comment (TheBigLee) [inline component/component/billing_cronjob.libsonnet:18]: This should probably be fixed

File: component/component/main.jsonnet [hunk 6]
```diff
@@ -112,6 +112,10 @@ local promQueryTemplate = importstr 'promql/appcat.promql';
 local promQueryWithLabel = std.strReplace(promQueryTemplate, '{{ORGLABEL}}', tenant.label);
 local promQuery = std.strReplace(promQueryWithLabel, '{{TENANT_REPLACE}}', tenant.replace);
 
+local meteringQueryCloud = importstr 'promql/metering_cloud.promql';
+local meteringQueryManagedRaw = importstr 'promql/metering_managed.promql';
+local meteringQueryManaged = std.strReplace(meteringQueryManagedRaw, '{{salesOrderID}}', params.salesOrderID);
```
Comment (TheBigLee) [inline component/component/main.jsonnet:117]:
```suggestion
local meteringQueryManaged = std.strReplace(meteringQueryManagedRaw, '{{salesOrder}}', params.salesOrderID);
```

File: component/component/main.jsonnet [hunk 7]
```diff
@@ -150,6 +154,64 @@ local promRule = kube._Object('monitoring.coreos.com/v1', 'PrometheusRule', 'app
...
+    groups: [
+      {
+        name: 'mock-org-info',
+        rules: [
+          {
+            expr: '1',
+            record: 'appuio_control_organization_info',
+            labels: {
+              name: 'awesomekorp',
+              sales_order_id: 'ST10120',
```
Comment (TheBigLee) [inline component/component/main.jsonnet:190]:
```suggestion
              sales_order: 'ST10120',
```

File: component/component/main.jsonnet [hunk 8]
```diff
@@ -150,6 +154,64 @@ local promRule = kube._Object('monitoring.coreos.com/v1', 'PrometheusRule', 'app
...
+            },
+          },
+          {
+            expr: '1',
+            record: 'appuio_control_organization_info',
+            labels: {
+              name: 'notvshn',
+              sales_order_id: 'invalid',
```
Comment (TheBigLee) [inline component/component/main.jsonnet:198]:
```suggestion
              sales_order: 'invalid',
```

File: component/component/main.jsonnet [hunk 9]
```diff
@@ -150,6 +154,64 @@ local promRule = kube._Object('monitoring.coreos.com/v1', 'PrometheusRule', 'app
...
+            },
+          },
+          {
+            expr: '1',
+            record: 'appui_managed_info',
+            labels: {
+              tenant_id: params.tenantID,
+              sales_order_id: 'ST10120',
```
Comment (TheBigLee) [inline component/component/main.jsonnet:206]:
```suggestion
              sales_order: 'ST10120',
```

File: component/component/promql/metering_cloud.promql [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  sum_over_time(
+    # Default appcat.vshn.io/sla to besteffort if it is not set
+    label_replace(
+        # Fetch all namespaces with label label_appuio_io_billing_name=~"appcat-.+"
+        (kube_namespace_labels{label_appuio_io_billing_name=~"appcat-.+"} *
+        on (namespace) group_right(label_appuio_io_organization,label_appcat_vshn_io_claim_namespace,label_appcat_vshn_io_sla,label_appcat_vshn_io_claim_name,label_appuio_io_billing_name,label_appuio_io_organization)
+        kube_pod_info{created_by_kind!="Job"}
+        # We join it with appuio_control_organization_info to map the organization to a sales_order_id
+        # This metric is not available on the cluster's prometheus, only on mimir, this is just included for completenes sake.
+        #* on(label_appuio_io_organization) group_left(sales_order_id) label_replace(appuio_control_organization_info, "label_appuio_io_organization", "$1", "name", "(.*)")
```
Comment (TheBigLee) [inline component/component/promql/metering_cloud.promql:12]:
```suggestion
        # We join it with appuio_control_organization_info to map the organization to a sales_order
        # This metric is not available on the cluster's prometheus, only on mimir, this is just included for completenes sake.
        #* on(label_appuio_io_organization) group_left(sales_order) label_replace(appuio_control_organization_info, "label_appuio_io_organization", "$1", "name", "(.*)")
```

File: component/component/promql/metering_managed.promql [hunk 11]
```diff
@@ -0,0 +1,25 @@
+sum by (label_appcat_vshn_io_claim_namespace,label_appcat_vshn_io_sla,label_appuio_io_billing_name,label_appcat_vshn_io_claim_name,sales_order_id,label_appuio_io_billing_name) (
```
Comment (TheBigLee) [inline component/component/promql/metering_managed.promql:1]:
```suggestion
sum by (label_appcat_vshn_io_claim_namespace,label_appcat_vshn_io_sla,label_appuio_io_billing_name,label_appcat_vshn_io_claim_name,sales_order,label_appuio_io_billing_name) (
```

File: component/component/promql/metering_managed.promql [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    label_replace(
+        # Fetch all namespaces with label label_appuio_io_billing_name=~"appcat-.+"
+        (
+          # We hardcode the tenant ID here, so that we have the label to do the join.
+          label_replace(
+              kube_namespace_labels{ label_appuio_io_billing_name=~"appcat-.+"} *
+              on (namespace) group_right(label_appcat_vshn_io_claim_namespace,label_appcat_vshn_io_sla, label_appuio_io_billing_name,label_appcat_vshn_io_claim_name,sales_order_id,label_appuio_io_billing_name)
+              kube_pod_info{created_by_kind!="Job"},
+            "sales_order_id",
+            "{{salesOrderID}}",
```
Comment (TheBigLee) [inline component/component/promql/metering_managed.promql:14]:
```suggestion
              on (namespace) group_right(label_appcat_vshn_io_claim_namespace,label_appcat_vshn_io_sla, label_appuio_io_billing_name,label_appcat_vshn_io_claim_name,sales_order,label_appuio_io_billing_name)
              kube_pod_info{created_by_kind!="Job"},
            "sales_order",
            "{{salesOrderID}}",
```

File: component/class/defaults.yml (hunk 2 above)
Comment (Kidswiss) [reply component/class/defaults.yml:78]: We probably won't need this one now, anyway.

File: component/component/billing.jsonnet (hunk 3 above)
Comment (Kidswiss) [reply component/component/billing.jsonnet:13]: I'm also not entirely sure, I've copied it from the appuio-reporting component.

But I think it just replaces any non-matching character with a `-`.

File: component/component/vshn_postgres.jsonnet [hunk 13]
```diff
@@ -23,6 +23,7 @@ local certificateSecretName = 'tls-certificate';
 local serviceNameLabelKey = 'appcat.vshn.io/servicename';
 local serviceNamespaceLabelKey = 'appcat.vshn.io/claim-namespace';
 local slaLabelKey = 'appcat.vshn.io/sla';
+local serviceCLaimNameLabelKey = 'appcat.vshn.io/claim-name';
```
Comment (TheBigLee) [inline component/component/vshn_postgres.jsonnet:26]: This variable is not used anywhere and also contains a typo (serviceC**L**aimNameLabelKey)

File: component/component/vshn_redis.jsonnet [hunk 14]
```diff
@@ -23,6 +23,7 @@ local clientCertificateSecretName = 'tls-client-certificate';
 
 local serviceNameLabelKey = 'appcat.vshn.io/servicename';
 local serviceNamespaceLabelKey = 'appcat.vshn.io/claim-namespace';
+local serviceCLaimNameLabelKey = 'appcat.vshn.io/claim-name';
```
Comment (TheBigLee) [inline component/component/vshn_redis.jsonnet:26]: Dito

File: component/tests/vshn.yml [hunk 15]
```diff
@@ -61,8 +71,8 @@ parameters:
             # grpcEndpoint: host.docker.internal:9443
             # proxyFunction: true
         postgres:
-          # grpcEndpoint: host.docker.internal:9443
-          # proxyFunction: true
+          grpcEndpoint: host.docker.internal:9443
+          proxyFunction: true
```
Comment (TheBigLee) [inline component/tests/vshn.yml:75]: This should be reverted before merging

File: component/component/vshn_postgres.jsonnet (hunk 13 above)
Comment (Kidswiss) [reply component/component/vshn_postgres.jsonnet:26]: Hmmm... I think i missed those when rebasing to master.

**Review by TheBigLee — APPROVED** [review]

File: package/main.yaml [hunk 16]
```diff
@@ -1,9 +1,10 @@
 applications:
   - appcat
   - crossplane
+  - appuio-reporting as appcat-reporting
 parameters:
   pkg.appcat:
-    componentVersion: v2.13.1
+    componentVersion: v2.14.0
```
Comment (TheBigLee) [inline package/main.yaml:7]: This is actually not necessary, as the Github Action will set the correct version automatically.

Comment (Kidswiss) [reply package/main.yaml:7]: Huh TIL, thanks.

**Pull request merged**